roles/tests-integration/tests/common/proxy/mod.rs

```diff
@@ -0,0 +1,489 @@
+            let mut send_original_frame_upstream = true;
+            for handler in handlers.iter_mut() {
+                if let Some(frame) = handler.on_message(&mut frame).await {
+                    send_original_frame_upstream = false;
+                    if send.send(frame).await.is_err() {
+                        return Err(ProxyError::UpstreamClosed);
+                    };
+                }
+            }
+            if send_original_frame_upstream && send.send(frame).await.is_err() {
+                return Err(ProxyError::UpstreamClosed);
+            };
+        }
+        Err(ProxyError::DownstreamClosed)
+    }
+
+    async fn recv_from_up_send_to_down(
+        mut recv: Receiver<MessageFrame>,
+        send: Sender<MessageFrame>,
+        mut handlers: Vec<MessageChannel>,
+    ) -> Result<(), ProxyError> {
+        while let Some(mut frame) = recv.recv().await {
+            let mut send_original_frame_upstream = true;
+            for handler in handlers.iter_mut() {
+                if let Some(frame) = handler.on_message(&mut frame).await {
+                    send_original_frame_upstream = false;
+                    if send.send(frame).await.is_err() {
+                        return Err(ProxyError::DownstreamClosed);
+                    };
+                }
+            }
+            if send_original_frame_upstream && send.send(frame).await.is_err() {
+                return Err(ProxyError::DownstreamClosed);
+            };
+        }
+        Err(ProxyError::UpstreamClosed)
+    }
+}
+
+#[derive(PartialEq, Clone, Debug)]
+pub enum Role {
+    Client,
+    Server,
+}
+
+impl std::fmt::Display for Role {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Role::Client => write!(f, "Client"),
+            Role::Server => write!(f, "Server"),
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+struct MessagesAggregator {
+    messages: Arc<Mutex<Vec<(PoolMessages<'static>, Role)>>>,
+}
+
+impl MessagesAggregator {
+    pub fn new() -> Self {
+        Self {
+            messages: Arc::new(Mutex::new(vec![])),
+        }
+    }
+
+    pub fn add_message(&self, message: PoolMessages<'static>, role: Role) {
+        self.messages
+            .safe_lock(|messages| {
+                let mut new_state = messages.clone();
+                new_state.push((message, role));
+                *messages = new_state
+            })
+            .unwrap();
+    }
+
+    pub fn contains(&self, message: ExpectMessage, role: Role) -> bool {
+        let messages = self
+            .messages
+            .safe_lock(|messages| messages.clone())
+            .unwrap();
+        let messages = messages
+            .iter()
+            .map(|(m, r)| (ExpectMessage::from(m.clone()), r.clone()));
+        for m in messages {
+            if m.0 == message && m.1 == role {
+                return true;
+            }
+        }
+        false
+    }
```

Comment:
what if the same message happens more than 1x?

how can we assert the second time, without getting a false positive coming from the first time this message ocurred?

Reply:
why do we want to keep a `Vec` of messages, instead of only keeping a record of the last message that was probed?

in case there's a meaningful justification to keeping a `Vec` of messages, like explained above, the `contains` function feels too generalistic and a potential footgun

assuming a `Vec` record is useful, instead of `contains`, I can imagine two other functions being useful here:
- `is_last_message(&self, message: ExpectMessage, role: Role) -> bool`
- `is_nth_message(&self, message: ExpectMessage, role: Role, n: usize) -> bool`

but again, I'm not too sure about how `is_nth_message` could be useful.

for most MG tests, we just need to assert against a specific message immediately after the moment where the message is expected to be sent/received

there's no real point in keeping a record for later analysis

so the `Vec` record feels like overkill, unless I'm missing something?